File: CPP/Solutions/2015/3/Year2015Day03.cpp
#include <utility>
#include <sstream>
#include <iostream>
#include <algorithm>
#include <numeric>
#include <set>
#include <tuple>

#include "Year2015Day03.h"

struct Point {
    int x;
    int y;
    Point() : x(0), y(0) {}
};

inline bool operator<(const Point& lhs, const Point& rhs) {
    return std::tie(lhs.x, lhs.y) < std::tie(rhs.x, rhs.y);
}

auto Year2015Day03::parse() {
    return this->input[0];
}

std::string Year2015Day03::part1() {
    std::string input = this->parse();
    auto santa = new Point();
    std::set<Point> points = {*santa};
    for (const auto &i: input) {
        switch (i){
            case '>':
                santa->y += 1;
                break;
            case '<':
                santa->y -= 1;
                break;
            case '^':
                santa->x += 1;
                break;
            case 'v':
                santa->x -= 1;
                break;
            default:
                break;
        }
        points.insert(*santa);
    }
    return std::to_string(points.size());
}

std::string Year2015Day03::part2() {
    auto santa = new Point();
    auto roboSanta = new Point();
    std::string input = this->parse();
    std::set<Point> points = {*santa};
    for(int i=0; i < input.size(); ++i) {
        Point* curSanta = i % 2 ? santa : roboSanta;
        switch (input[i]){
            case '>':
                curSanta->y += 1;
                break;
            case '<':
                curSanta->y -= 1;
                break;
            case '^':
                curSanta->x += 1;
                break;
            case 'v':
                curSanta->x -= 1;
                break;
            default:
                break;
        }
        points.insert(*curSanta);
    }
    return std::to_string(points.size());
}

Year2015Day03::Year2015Day03(std::vector<std::string> lines) {
    this->input = std::move(lines);
}

File: CSharp/Solutions/2015/3/Year2015Day03.cs
using System.Collections.Generic;
using System.Linq;
using AdventOfCode.Common;

namespace AdventOfCode.Solutions._2015._3
{
	// ReSharper disable once UnusedMember.Global
	public class Year2015Day03 : ISolution
	{
		public object Part1(IEnumerable<string> input)
		{
			var directions = input.First();
			var location = new {X = 0, Y = 0};
			var points = new HashSet<object>{location};
			foreach (var direction in directions)
			{
				if (direction == '>')
					location = new {X = location.X + 1, location.Y};
				if (direction == '<')
					location = new {X = location.X - 1, location.Y};
				if (direction == '^')
					location = new {location.X, Y = location.Y + 1};
				if (direction == 'v')
					location = new {location.X, Y = location.Y - 1};
				points.Add(location);
			}
			return points.Count.ToString();
		}

		public object Part2(IEnumerable<string> input)
		{
			var directions = input.First();
			var santa = new {X = 0, Y = 0};
			var roboSanta = new {X = 0, Y = 0};
			var santaTurn = true;
			var points = new HashSet<object>{santa};
			foreach (var direction in directions)
			{
				var location = santaTurn ? santa : roboSanta;
				if (direction == '>')
					location = new {X = location.X + 1, location.Y};
				if (direction == '<')
					location = new {X = location.X - 1, location.Y};
				if (direction == '^')
					location = new {location.X, Y = location.Y + 1};
				if (direction == 'v')
					location = new {location.X, Y = location.Y - 1};
				if (santaTurn)
					santa = location;
				else
					roboSanta = location;
				santaTurn = !santaTurn;
				points.Add(location);
			}
			return points.Count.ToString();
		}
	}
}